### PR TITLE
TST, BUILD: add a gcc 4.8 run on ubuntu 18.04

### PR DESCRIPTION
**Pull request by mattip**
Add a gcc 4.8 build to Azure, with the goal of properly adding a std=c99 flag

### CODE REVIEW

File: numpy/distutils/system_info.py [hunk 1]
```diff
@@ -1580,7 +1580,7 @@ def get_atlas_version(**config):
             log.info('Status: %d', s)
             log.info('Output: %s', o)
 
-    if atlas_version == '3.2.1_pre3.3.6':
+    elif atlas_version == '3.2.1_pre3.3.6':
```
Comment (mattip) [inline numpy/distutils/system_info.py]: Before this change, building locally with a broken compiler would result in the call to `_c_string_literal` below erroring out 
```
  File "/home/matti/pypy_stuff/numpy/numpy/distutils/system_info.py", line 1587, in get_atlas_version
    'ATLAS_INFO', _c_string_literal(atlas_version))
  File "/home/matti/pypy_stuff/numpy/numpy/distutils/system_info.py", line 177, in _c_string_literal
    s = s.replace('\\', r'\\')
AttributeError: 'NoneType' object has no attribute 'replace'

```

**Comment by mattip** [comment]
It seems the place to add the `std=c99` flag is in `numpy/distutils/ccompiler.CCompiler_customize`. This should be called only a few times, but is called ~60 times by the system_info lapack discovery phase since it does not cache the compiler. So I would like to try to compile some c99 code and see if it fails, but it seems excessive to compile it each time.

**Comment by rgommers** [comment]
`CCompiler_customize` seems like a reasonable place to add this indeed. It should always be called (we're not enforcing that, although we should - it leads to hard to detect bugs otherwise).

> So I would like to try to compile some c99 code and see if it fails, but it seems excessive to compile it each time.

Just compile it once and cache the result?

**Comment by charris** [comment]
> So I would like to try to compile some c99 code and see if it fails,

IIRC, the bit that caused problems was declarations in for loops:
```
for(int i; i < 3; ++i);
```
Declaration after statement was another, but I think that passed without c99,

**Comment by mattip** [comment]
I could not find a way to quiet the compiler output when calling `self.compile` inside `CCompiler_customize`. I think adding the `-std=c99` flag will always be safe when the compiler name is `gcc`

File: azure-pipelines.yml [hunk 2]
```diff
@@ -64,7 +64,7 @@ stages:
       # the docs even though i.e., numba uses another in their
       # azure config for mac os -- Microsoft has indicated
       # they will patch this issue
-      vmImage: macOS-10.13
+      vmImage: macOS-10.14
```
Comment (mattip) [inline azure-pipelines.yml]: This change is enough to fix the macOS failure

File: doc/release/upcoming_changes/14771.improvement.rst [hunk 3]
```diff
@@ -0,0 +1,6 @@
+``std=c99`` added if compiler is named ``gcc``
+----------------------------------------------
+
```
Comment (charris) [inline doc/release/upcoming_changes/14771.improvement.rst]: I don't think the blank line is needed here. I edit them out of the final notes.

File: numpy/distutils/ccompiler.py [hunk 4]
```diff
@@ -532,6 +532,11 @@ def CCompiler_customize(self, dist, need_cxx=0):
                                       'g++' in self.compiler[0] or
                                       'clang' in self.compiler[0]):
         self._auto_depends = True
+        if 'gcc' in self.compiler[0]:
+            # add std=c99 flag for gcc < 5
```
Comment (charris) [inline numpy/distutils/ccompiler.py]: This comment no longer seems valid.

**Comment by charris** [comment]
Seems to work, just a couple of nits. Might enable travis as you touched those tests also.

**Comment by mattip** [comment]
Fixed the review comments. Travis is now running on this PR

**Comment by charris** [comment]
Any idea what is going on with the "abandoned" failing tests?

**Comment by charris** [comment]
I don't think the failing tests are relevant, but they also should not be failing. I'm going to close/reopen to see if that changes anything.

**Comment by charris** [comment]
Arrggghhhh, those staged azure tests really slow things down, the stages run serially rather than in parallel so that it takes twice as long. What a horrible waste of time :)

**Comment by charris** [comment]
Thanks Matti.

**Pull request merged**